2016/notes-on-testing-with-jest.md

```diff
@@ -0,0 +1,280 @@
+
+## FAQs
+
+The FAQs are categorised by the facet of testing that they cover. These may also apply to any testing in general.
+
+### Fundamentally speaking
+
+__Why should we move to `Jest` when we already have a working setup?__
+
+A move to Jest would mean a somewhat change in paradigm and would involve team effort. We justified the effort with the following reasons
+
+  - Jest is way faster than our current setup. It runs test suites in a parallel and the community is focused on improving developer experience through performance.
+  - I really liked the interactive CLI that comes with it. In `watch` mode, Jest runs only the tests associated with your code changes. But it allows you to interact with the tests by providing functionality to update snapshots, search for a particular suite, run all tests if you want to.
+  - [Great documentation](https://facebook.github.io/jest/docs/getting-started.html)
+  - Snapshot testing for components. It is kind of tailor made for React apps.
+  - Easy configuration and great defaults. Running `jest` alone would be enough to run test files. Mocking is not a marriage of complexity and confusion.
+  - Existing tests could be [easily migrated](https://facebook.github.io/jest/docs/migration-guide.html#content)
+
+__What about testing on the browser?__
+
+Yes, the earlier setup opened a chromium browser and tested our components. It was pretty cool actually. But, it was one of the things that made executing tests slow for us. Being priviledged enough to have manual QA, we decided to forego this functionality.
+
+For more reasoning, you should [read this](https://github.com/facebook/jest/issues/139#issuecomment-229277654) by [@cpojer](https://github.com/cpojer) on why Facebook doesn't need them.
+Also, we are open to using [React storybook](https://getstorybook.io/) in the future, which I think has potential for browser testing.
+
+__What is Snapshot testing? Will it really help?__
+
+Snapshot testing, is a procedure in which, Jest generates readable snapshots of your component's view. A snapshot is basically a text file with xml, which acts like the source of your view, at that point in your testcase. Different snapshots can be created with different props to your components.
+
+Snapshots can be saved after simulating events like `click` and `hover`. The point would be to save the expected view of your component which is a function of some mock data. If in the future, a developer changes your component, Jest would create a new snapshot of the component and compare it with the saved snapshot. If there is a change in view, it will alert the developer and she will have to tell Jest to save this new snapshot. Since the snapshot is committed to version control, it will show up in a code review and the changes can be approved accordingly.
+While we still have reservations of, if its actually feasible to review large changes in the snapshot, I definitely think this enriches the content of a code review.
```

Comment:
It also may be worth calling out that the integration we did is pretty substantial. For example Jest keeps track of outdated snapshots, obsolete snapshots and obsolete snapshot files. If you delete a test file but not the associated snapshot file, Jest will fail the test run and vice versa. It covers all of the cases to make sure the project stays consistent.